MSAL/src-internal/ios/MSALLegacySharedAccount.m

```diff
@@ -0,0 +1,211 @@
+{
+    self = [super init];
+    
+    if (self)
+    {
+        _jsonDictionary = jsonDictionary;
+        _accountType = [jsonDictionary msidStringObjectForKey:@"type"];
+        _accountIdentifier = [jsonDictionary msidStringObjectForKey:@"id"];
+        
+        if ([NSString msidIsStringNilOrBlank:_accountType]
+            || [NSString msidIsStringNilOrBlank:_accountIdentifier])
+        {
+            MSID_LOG_WITH_CTX(MSIDLogLevelError, nil, @"Missing account type or identifier (account type = %@, account identifier = %@)", _accountType, _accountIdentifier);
+            
+            if (error)
+            {
+                *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInternal, @"Unexpected shared account found without type or identifier", nil, nil, nil, nil, nil);
+            }
+            
+            return nil;
+        }
+        
+        _signinStatusDictionary = [jsonDictionary msidObjectForKey:@"signInStatus" ofClass:[NSDictionary class]];
+        MSID_LOG_WITH_CTX_PII(MSIDLogLevelInfo, nil, @"Created sign in status dictionary %@", MSID_PII_LOG_MASKABLE(_signinStatusDictionary));
+    }
+    
+    return self;
+}
+
+- (instancetype)initWithMSALAccount:(id<MSALAccount>)account
+                      accountClaims:(NSDictionary *)claims
+                    applicationName:(NSString *)appName
+                     accountVersion:(MSALLegacySharedAccountVersion)accountVersion
+                              error:(NSError **)error
+{
+    if (accountVersion == MSALLegacySharedAccountVersionV1)
+    {
+        return nil;
+    }
+    
+    if (!account)
+    {
+        if (error)
+        {
+            *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInternal, @"Unexpected parameter - no account", nil, nil, nil, nil, nil);
+        }
+        
+        return nil;
+    }
+    
+    NSString *appBundleId = [[NSBundle mainBundle] bundleIdentifier];
+    
+    NSMutableDictionary *jsonDictionary = [NSMutableDictionary new];
+    jsonDictionary[@"id"] = [[NSUUID UUID] UUIDString];
+    jsonDictionary[@"environment"] = @"PROD";
+    
+    if (accountVersion == MSALLegacySharedAccountVersionV3)
+    {
+        jsonDictionary[@"originAppId"] = appBundleId;
+    }
+    
+    jsonDictionary[@"signInStatus"] = @{appBundleId : @"SignedIn"};
+    jsonDictionary[@"username"] = account.username;
+    jsonDictionary[@"additionalProperties"] = @{@"createdBy": appName};
+    [jsonDictionary addEntriesFromDictionary:[self claimsFromMSALAccount:account claims:claims]];
+    return [self initWithJSONDictionary:jsonDictionary error:error];
+}
+
+#pragma mark - Match
+
+- (BOOL)matchesParameters:(MSALAccountEnumerationParameters *)parameters
+{
+    if (parameters.needsAssociatedRefreshToken)
+    {
+        NSString *appIdentifier = [[NSBundle mainBundle] bundleIdentifier];
+        NSString *signinStatus = _signinStatusDictionary[appIdentifier];
+        
+        if (![signinStatus isEqualToString:@"SignedIn"])
```

Comment:
nit:Should we check if signInStatus is not nil.

Reply:
if it's nil, it will be fine. but we should check that it's of NSString type.